solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java

```diff
@@ -49,411 +58,673 @@
+        activePushers--;
+      } else if (currentStage == SharedCoreStage.INDEXING_BATCH_FINISHED) {
         activeIndexers--;
       }
 
-      // making sure no other activity takes place during pull
-      if (prevStage == SharedCoreStage.BlobPullStarted) {
-        assertTrue("Pull critical section breached, currentStage=" + p, prevThreadId==currentThreadId && currentStage == SharedCoreStage.BlobPullFinished);
-      }
-
-      // making sure indexing is not disrupted by a pull from blob
-      assertFalse("Indexing breached by a pull, currentStage=" + p,
-          activeIndexers > 0 &&
-              (currentStage == SharedCoreStage.BlobPullStarted || currentStage == SharedCoreStage.BlobPullFinished));
+      // making sure no other activity (including another pull) takes place during pull
+      assertFalse("Pull and indexing are interleaved, coreName=" + coreName + " currentStage=" + s, activePullers > 1 || (activePullers > 0 && (activeIndexers > 0 || activePushers > 0)));
 
       // making sure push to blob are not disrupted by another push to blob
-      assertFalse("Blob push breached by another blob push,  currentStage=" + p, activeBlobPushers > 1);
+      assertFalse("Blob push breached by another blob push, coreName=" + coreName + " currentStage=" + s, activePushers > 1);
+    }
+  }
 
-      prevThreadId = currentThreadId;
-      prevStage = currentStage;
+  /**
+   * Adds a new replica.
+   */
+  private Replica addReplica() throws Exception {
+    List<String> existingReplicas = getCollection().getSlice(SHARD_NAME).getReplicas().stream().map(r -> r.getName()).collect(Collectors.toList());
+    // add another replica
+    assertTrue(CollectionAdminRequest.addReplicaToShard(COLLECTION_NAME, SHARD_NAME, Replica.Type.SHARED)
+        .process(cluster.getSolrClient()).isSuccess());
+    // Verify that new replica is created
+    waitForState("Timed-out waiting for new replica to be created", COLLECTION_NAME, clusterShape(1, existingReplicas.size() + 1));
+
+    Replica newReplica = null;
+
+    for (Replica r : getCollection().getSlice(SHARD_NAME).getReplicas()) {
+      if (!existingReplicas.contains(r.getName())) {
+        newReplica = r;
+        break;
+      }
     }
+
+    assertNotNull("Could not find new replica", newReplica);
+
+    return newReplica;
   }
 
-  private void configureTestSharedConcurrencyControllerForNode(JettySolrRunner runner, List<String> progress) {
-    SharedCoreConcurrencyController concurrencyController = new SharedCoreConcurrencyController(runner.getCoreContainer()) {
-      Object recorderLock = new Object();
-      @Override
-      public void recordState(String collectionName, String shardName, String coreName, SharedCoreStage stage) {
-        super.recordState(collectionName, shardName, coreName, stage);
+  /**
+   * Directly query the {@code replica} and verify that all the docs indexed(after accounting for deletions) are found.
```

Comment:
Might be worth adding a note that this is testing an empty replica hydrating after the concurrent tests